### PR TITLE
Workaround a bug in envoy

### PR DESCRIPTION
**Pull request by marcoferrer**
This is a fix to bring baseplate go in sync with baseplate py.

### CODE REVIEW

File: thriftbp/headers.go [hunk 1]
```diff
@@ -44,3 +45,27 @@ func AddClientHeader(ctx context.Context, key, value string) context.Context {
 	headers = append(headers, key)
 	return thrift.SetWriteHeaderList(ctx, headers)
 }
+
+// Header gets the value of a thrift header by key
+//
+// If the value is not present we fall back to a case-insensitive check
+func Header(ctx context.Context, key string) (v string, ok bool) {
```
Comment (fishy) [inline thriftbp/headers.go:52]: I don't think we want to export this. if a service code need to do the same thing they can just duplicate this function.

File: thriftbp/headers.go [hunk 2]
```diff
@@ -44,3 +45,27 @@ func AddClientHeader(ctx context.Context, key, value string) context.Context {
...
+
+	// Option 1.
+	v, ok = thrift.GetHeader(ctx, key)
+	if !ok {
+		v, ok = thrift.GetHeader(ctx, strings.ToLower(key))
```
Comment (fishy) [inline thriftbp/headers.go:57]: if you are certain that what envoy does is just lowercase the key, then this is enough and we won't need option 2.

**Review by fishy — CHANGES_REQUESTED** [review]

File: thriftbp/headers.go [hunk 3]
```diff
@@ -44,3 +45,17 @@ func AddClientHeader(ctx context.Context, key, value string) context.Context {
 	headers = append(headers, key)
 	return thrift.SetWriteHeaderList(ctx, headers)
 }
+
+// header gets the value of a thrift header by key
+//
+// If the value is not present we fall back to a case-insensitive check
```
Comment (fishy) [inline thriftbp/headers.go:51]: s/fall back to a case-insensitive check/fall back to all lowercase check to workaround a bug in envoy/

**Review by fishy — COMMENTED** [review]
the code lgtm, but before we can merge it:

1. The PR title should be something like "Workaround a bug in envoy" as this is not a bug in baseplate.go
2. We need to file a bug on envoy, and link the bug here, so we could revert this change in the future when the bug is fixed in envoy.

**Comment by fishy** [comment]
https://github.com/envoyproxy/envoy/issues/20595

File: thriftbp/headers.go [hunk 4]
```diff
@@ -44,3 +45,17 @@ func AddClientHeader(ctx context.Context, key, value string) context.Context {
...
+func header(ctx context.Context, key string) (v string, ok bool) {
+	v, ok = thrift.GetHeader(ctx, key)
+	if !ok {
+		// We fall back to a case-insensitive check so that we can have
```
Comment (kylelemons) [inline thriftbp/headers.go:55]: This is incorrect -- it's not a case-insensitive check, it's a lower-case check.

I think you should say
```
We fall back to checking the key that an envoy proxy will send, which is a lower-case
version of the thrift key.

For details on this workaround, see:
* <link to this PR or whatever>
* <link to the envoyproxy/envoy bug>
```

**Review by kylelemons — CHANGES_REQUESTED** [review]

File: thriftbp/headers.go [hunk 5]
```diff
@@ -44,3 +45,17 @@ func AddClientHeader(ctx context.Context, key, value string) context.Context {
...
+		// functional parity with baseplate-py
```
Comment (marcoferrer) [inline thriftbp/headers.go:56]:
```suggestion
		// We fall back to checking the key that an envoy proxy will send, which is a lower-case
		// version of the thrift key.
		// 
		// 
		// For details on this workaround, see:
		// 	* https://github.com/reddit/baseplate.py/blob/0d189bae9bc15459b8d9d62856f31e45cbfcdb1a/baseplate/frameworks/thrift/__init__.py#L111	
		// 	* https://github.com/envoyproxy/envoy/issues/20595
		// 	* https://github.com/reddit/baseplate.go/pull/500
```

File: thriftbp/headers.go [hunk 6]
```diff
@@ -44,3 +45,23 @@ func AddClientHeader(ctx context.Context, key, value string) context.Context {
...
+// If the value is not present we fall back to a case-insensitive check
+func header(ctx context.Context, key string) (v string, ok bool) {
+	v, ok = thrift.GetHeader(ctx, key)
+	if !ok {
+		// We fall back to checking the key that an envoy proxy will send, which is a lower-case
+		// version of the thrift key.
+		// 
+		// 
+		// For details on this workaround, see:
+		// 	* https://github.com/reddit/baseplate.py/blob/0d189bae9bc15459b8d9d62856f31e45cbfcdb1a/baseplate/frameworks/thrift/__init__.py#L111
```
Comment (marcoferrer) [inline thriftbp/headers.go:60]:
```suggestion
		// 	* https://github.com/reddit/baseplate.py/commit/daf1c5e32dcc019470d75f3940006bb2f7b9855e	
```

File: thriftbp/headers.go [hunk 7]
```diff
@@ -44,3 +45,23 @@ func AddClientHeader(ctx context.Context, key, value string) context.Context {
 	headers = append(headers, key)
 	return thrift.SetWriteHeaderList(ctx, headers)
 }
+
+// header gets the value of a thrift header by key
+//
+// If the value is not present we fall back to a case-insensitive check
```
Comment (marcoferrer) [inline thriftbp/headers.go:51]:
```suggestion
// If the value is not present we fall back to all lowercase check to workaround a bug in envoy
```

**Review by fishy — APPROVED** [review]

**Review by kylelemons — APPROVED** [review]

**Review by pacejackson — APPROVED** [review]

**Comment by fishy** [comment]
you have extra trailing spaces: https://github.com/reddit/baseplate.go/runs/5775116141?check_suite_focus=true

**Pull request merged**